### ISSUE TITLE
Update/Remove babel-core in a new major version

### ISSUE THREAD

**Issue by sokra** [facebook/jscodeshift#229]
Currently jscodeshift depends on an old version of babel-core, which depends on `minimatch@2` which
* Reports a security warning on `npm install` pointing to a security issue
* Reports errors for `nsp` which checks for security warnings

Because of these two problems I'm unwilling to use this package. These warnings/errors are transitive and I don't want my users to see security issues reported. Note that this is not a real security issue for jscodeshift, because it's not using the affected babel options. Nevertheless I don't want to tell every user to ignore these warnings, because it affects the trust in security warnings!

In another issue you pointed out that removing/updating babel would be a backward incompatible change:

see https://github.com/facebook/jscodeshift/pull/198#issuecomment-304329657

> Yeah, this is not something we can do because we need babel v5 for backwards compatibility. Maybe if we release a new breaking version we could remove babel v5 support, but not at the moment.

So I propose to increase the major version and release a breaking change.

Do you really want a security warning reported on every npm install of your package? This affects the trust in security warnings! This trust is important for the community! Especially a package under the facebook namespace should be an example.

**Comment by slorber** [comment]
@sokra as mentionned on Twitter, I was able to run successfully jscodeshift on babel6 by backporting the `babel.parse` function from babel5 to babel6. 

This function has been deleted in babel6 and in babel5 it was just a wrapper around Babylon parser with some preset.

Check source code here for backport code:
https://github.com/slorber/react-reboot/blob/master/src/lib/react-reboot.js#L23

**Comment by TheLarkInn** [comment]
@fkling Is there anything we can help or do to unlock this? Happy to help in any way. This is a top 5 priority item for us to unlock merging webpack-cli into webpack which is a key integration for webpack 4. 

cc @cpojer incase notification aren't making it to Felix (which they do with me most times).

**Comment by ematipico** [comment]
@slorber I don't think it's a matter of make it working (thank you very much for the workaround provided), but it's more like the feedback that the library is going to provide during the installation. 

How would you feel, as end-user, to see that an OSS library from an important company shows a warning related to a security issue? I thinks that's the message of the argument. Security patches have to provided.

**Comment by slorber** [comment]
@sokra @ematipico to me it is not really clear, is one of your libraries importing jscodeshift? or do you worry about direct users of jscodeshift?

@ematipico the thing is, if we can make it work with babel6, then you can upgrade safely and won't get the warning again.

I even think that the babel-core might be removed completely and only keep the babylon parsing, but I'll have to check that to confirm

**Comment by slorber** [comment]
I've made a pull request that should work fine in a retrocompatible way, without depending on babel-core@^5

=> https://github.com/facebook/jscodeshift/pull/230

**Comment by fkling** [comment]
@hzoo, what’s the timeline for babylon 7?

@slorber: not sure this is going to work. Babylon 6 still produces a different AST than babel 5.

Overall I don’t have a problem dropping support for babel5. I’d hope that most codemods would work with flow at least (which is more estree compatible than babylon 6).

If most codemods are distributed as “standalone” packages then that’s less of a problem since they’d probably require a specific version of jscodeshift.

However, I’m on parental leave until January and cannot promise I find time to do this soon.

**Comment by slorber** [comment]
@fkling the Babylon parser has an "estree" plugin to be retrocompatible with babel5 format. I've added that plugin in the config and already used that config to migrate my own codebase successfully.

It's also the setup that I use for [React-reboot](react-reboot.now.sh) which runs JsCodeshift codemods on a node server with an API. 

Regarding Flow Babylon plugin, I had to remove it from config to make tests pass because it seems tests are asserting that default parser is unable to parse flow, so I kept this behavior for retrocompatibility:

```javascript
    it('errors if new flow type code is parsed with babel v5', done => {
      const transformPath = createTransformWith(
        'api.jscodeshift(fileInfo.source); return "changed";'
      );
      const sourcePath = getSourceFile();
      const emitter = worker([transformPath]);

      emitter.send({files: [sourcePath]});
      emitter.once('message', (data) => {
        expect(data.status).toBe('error');
        expect(data.msg).toMatch('SyntaxError');
        done();
      });
    });
```

**Comment by TheLarkInn** [comment]
@fkling are you the only owner here at FB? I certainly don't expect you to tackle this (enjoy the new baby, congrats!). However I'd still like to put a bit of pressure where possible if there is someone with bandwidth. cc @cpojer <3 

I only stress as its really important for webpack-cli getting into webpack :-D

**Comment by cpojer** [comment]
There is no owner at FB for jscodeshift at this time and we haven't had one for a while. @fkling's dedication to the project was mostly outside of normal work hours (Felix, correct me if I'm wrong). I'm afraid you won't be able to find anyone here at the moment who'll be able to take this on but we are happy to help with pull request reviews and releasing new versions. I'm sorry if that's inconvenient but we are focused on a number of other things currently. This is a great opportunity for people in the community to step up and help us out.

**Comment by slorber** [comment]
@cpojer there is already a PR, what else can be done by the community to help so that it can be merged and an alpha release be published?

I suppose if the tests of this repo + react-codemods + js-codemods are all green, that is would be sufficient to make a new release no? (only run tests of this repo for now)

**Comment by cpojer** [comment]
I didn't realize there was a PR, I was just trying to manage expectations. I do not currently have access to make jscodeshift releases. The PR sent looks good to me but yes, I'd like to see react-codemod and js-codemod be green with those changes as well. I will try to see if Felix is available to give me access, assuming he is on board with the changes.

**Comment by ematipico** [comment]
This was the PR: https://github.com/facebook/jscodeshift/pull/198

Are there any other changes that needs to be done in order to have a beta release?

**Comment by slorber** [comment]
@ematipico you can't just upgrade babel like that because babel6 is not retrocompatible with babel5 (it does not have the parse() method anymore, so you need to call Babylon directly with correct setup to produce an AST compatible with babel5)

My PR tries to keep this retrocompatibility: https://github.com/facebook/jscodeshift/pull/230

This permits to ensure that normally all existing codemods would still run fine against new version of jscodeshift

**Comment by ematipico** [comment]
@slorber Didn't see THAT PR, sorry!

**Comment by fkling** [comment]
@TheLarkInn: Thank you!
@cpojer: Yep, mostly outside work hours... add you to the npm package.
@slorber: I thought the plugin was only available for babylon v7? Was it backported? Either way, if tests pass that's fine for me 😉 

> tests are asserting that default parser is unable to parse flow

Probably only to verify that passing `--parser=flow` actually selects the flow parser. It's probably better to enable the plugin and remove those other tests (if not too complicated).

**Comment by cpojer** [comment]
I merged the PR associated with this issue and published jscodeshift 0.4.0. Let me know if that works or doesn't work for all of you. Thanks @fkling, hope you are having an amazing time.

**Comment by slorber** [comment]
Don't know if this was backported but it's in plugin list :)

great to see this solved :) thanks
